README.md
# zbx-webhook-proxy

Accepts requests on a configurable set of URLs then forwards the GET query parameters or POSTed JSON body to Zabbix Server (as a JSON value), with optional [JMESPath](https://jmespath.org) transformation.

Use with Zabbix 4.0 or later and [Trapper Items](https://www.zabbix.com/documentation/6.0/en/manual/config/items/itemtypes/trapper) of type `log`, then configure further processing of the JSON body in Zabbix using [Dependent Items](https://www.zabbix.com/documentation/6.0/en/manual/config/items/itemtypes/dependent_items) and [JSON preprocessing](https://www.zabbix.com/documentation/6.0/en/manual/config/items/preprocessing/jsonpath_functionality).

## Built With

* [warp](https://crates.io/crates/warp) - A super-easy, composable, web server framework for warp speeds 
* [zbx-sender](https://crates.io/crates/zbx_sender) - Modern Rust implementation of Zabbix Sender Client
* [trust-dns-resolver](https://crates.io/crates/trust-dns-resolver) - Used for reverse DNS resolution not supported by `std::net`
* [clap](https://crates.io/crates/clap) - Parse command line arguments by defining a struct 
* [fern](https://crates.io/crates/fern) - “Simple, efficent logging”; manages console logs and optional access log file
* [serde-json](https://crates.io/crates/serde_json) - for `json!()` happiness

## Version History

See changes in [CHANGELOG.md](CHANGELOG.md)

## Building

### Prerequisites

Rust 2021 edition

### Installation
1. `cargo build --release`
2. Copy `target/release/zbx-webhook-proxy` to any executable location and configure startup

### TODO
- Startup scripts and SystemD unit files
- `make install`

## Configuration

### `config.toml` file

See [`example_config.toml`](example_config.toml) for full usage

**Minimal configuration**
```toml
zabbix_server = "zabbix.example.com"

[[route]]
path = "/hook"
item_host = "Monitored Host"
item_key = "event.json"
```

### Command Line
*All command-line flags are optional but can be specified to override the configuration file or defaults*

#### Flags with no default value
*MUST be specified on the command line or in the configuration file*
- `-z/--zabbix-server`: Address of Zabbix Server or Proxy<br>

#### Flags with a default value
- `-c/--config`: Path to `config.toml` (default `/etc/zbx-webhook-proxy.toml`)
- `-p/--zabbix-port`: Zabbix trapper port (default: 10051)
- `-l/--listen`: Proxy HTTP listening port (default: 3030)

#### Flags that enable additional features
- `--access-log`: Path to a log file that will contain requests in Apache Common log format

#### Flags that enable features only available from the command line
- `--test-mode`: Accept webhooks and print values that would be sent to Zabbix Server without actually sending
- `-v/--verbose`: Increase logging to the console for troubleshooting purposes

## Versioning

I use [SemVer](http://semver.org/) for versioning. For the versions available, see the [tags on this repository](https://github.com/overhacked/zbx-webhook-proxy/tags). 

## Authors

* **Ross Williams** - 🐙[GitHub](https://github.com/overhacked) &mdash; 🐦[Twitter](https://twitter.com/overhacked)

## License

This project is licensed under the Modified BSD License - see the [LICENSE](LICENSE) file for details


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.3.0] 2022-07-29
### Changed
- Updated to `zbx-sender-rs` 0.3.0, to address some vulnerabilities in
  transitive dependencies of that crate
- Switched from `logging` crate to `tracing` crate for console
  output and HTTP access logging. HTTP access log format remains
  the same; only the internal implementation has changed
- Console debugging output is richer and can be in color, showing
  message levels and module location of events
### Added
- Documented `"*"` value for `item_key` configuration, which expands each
  top-level key in the JSON object of a request to a separate Item in Zabbix.
  See [example configuration](example_config.yaml) for details.

## [0.2.0] 2022-07-18
### Added
- TOML configuration format
- Support for multiple URIs
- [JMESPath](https://jmespath.org) filtering and transformation of request data
- Dynamic hostname based on request data or reverse DNS

## [0.1.0-alpha] 2019-10-20
### First Release
- Support only GET parameters
- Send to Zabbix as a JSON hash, one key per parameter
- Access log in Apache Common format to a separate file


src/main.rs
mod config;
mod handlers;
mod filters;
mod logging;
mod zabbix;

use simple_eyre::eyre::Result;
use tracing::{info, error};
pub(crate) use serde_json::{json, Value as JsonValue};
use std::io;
use std::sync::Arc;
use warp::Filter;
use thiserror::Error;
use trust_dns_resolver::{error::ResolveError, TokioAsyncResolver as AsyncResolver};

use crate::config::Config;
use crate::zabbix::ZabbixLogger;
pub use crate::zabbix::ZabbixItemValue;

#[tokio::main]
async fn main() -> Result<()> {
    simple_eyre::install()?;
    let config = Config::load()?;

    let _log_guard = logging::setup(
        config.log_level,
        &config.access_log_path,
    )?;

    if let Some(path) = &config.access_log_path {
        info!("Logging HTTP requests to {}", path.display());
    }

    // The Zabbix connector must live inside Arc because it is not Clone
    let zabbix = Arc::new(
        ZabbixLogger::new(config.zabbix_server.expect("Config invariant"), config.zabbix_port)
    );
    // The AsyncResolver IS Clone but should live inside Arc to allow
    // efficient resolver caching, otherwise every resolver.clone()
    // copies the entire cache
    let resolver = Arc::new(
        AsyncResolver::tokio_from_system_conf()?
    );

    let mut route_filters = None;
    for route in config.routes {
        info!("Listening to requests on path `{}`", &route.path);
        let route_filter = filters::make_path_filter(route.path.as_str())
            .map(move || route.clone());

        match route_filters.take() {
            None => {
                route_filters = Some(route_filter.boxed());
            },
            Some(f) => {
                route_filters = Some(f.or(route_filter).unify().boxed());
            },
        }
    }

    let context = handlers::AppContext {
        zabbix: Arc::clone(&zabbix),
        resolver: Arc::clone(&resolver),
        test_mode: config.test_mode,
    };
    let routes = route_filters.expect("Config invariant")
        .and(filters::with_context(context))
        .and(warp::addr::remote())
        .and(filters::get().or(filters::post()).unify())
        .and_then(handlers::handle_request)
        .recover(handlers::handle_errors)
        .with(warp::trace(logging::warp_trace));

    warp::serve(routes).run(config.listen).await;
    Ok(())
}

#[derive(Error, Debug)]
enum AppError {
    #[error(transparent)]
    LoggingInit(#[from] tracing_subscriber::util::TryInitError),
    #[error(transparent)]
    Io(#[from] io::Error),
    #[error(transparent)]
    ResolverInit(#[from] ResolveError),
    #[error("Failed to load configuration file at `{path}`")]
    ConfigLoadError {
        path: String,
        source: io::Error
    },
    #[error(transparent)]
    ConfigSyntaxError(#[from] toml::de::Error),
    #[error(
        "Required setting `{0}` not present \
        in configuration file or on command line"
    )]
    ConfigMissingRequired(String),
}


src/zabbix.rs
use std::{borrow::Borrow, fmt::Display};

use serde::Serialize;
use tracing::{info, debug};

pub struct ZabbixLogger {
    sender: zbx_sender::Sender,
}

impl ZabbixLogger {
    pub fn new(server: impl Into<String>, port: u16) -> Self {
        let server = server.into();

        info!(
            "Logging to Zabbix Server at {}:{}",
            server, port
        );
        Self {
            sender: zbx_sender::Sender::new(server, port),
        }
    }

    pub fn log_many(&self,
        host: &str,
        values: impl IntoIterator<Item = impl Borrow<ZabbixItemValue>>
    )
        -> zbx_sender::Result<zbx_sender::Response>
    {
        let values: Vec<zbx_sender::SendValue> = values.into_iter()
            .map(|i| {
                let i = i.borrow();
                (host, i.key.as_str(), i.value.as_str(),).into()
            })
            .collect();

        debug!("sending to Zabbix `{:?}`", values);
        self.sender.send(values)
    }

    // fn log(&self, host: &str, key: &str, value: &str) -> zbx_sender::Result<zbx_sender::Response> {
    //     self.log_many(host, [(key, value,)])
    // }
}

#[derive(Debug, Serialize)]
pub struct ZabbixItemValue {
    pub key: String,
    pub value: String,
}

impl<S> From<(S, S,)> for ZabbixItemValue
    where S: Into<String>
{
    fn from((key, value,): (S, S,)) -> Self {
        Self {
            key: key.into(),
            value: value.into(),
        }
    }
}

#[derive(Debug, Serialize)]
pub(crate) struct TestZabbixValue<'a> {
    pub host: String,
    #[serde(flatten)]
    pub value: &'a ZabbixItemValue,
}

#[derive(Debug, Serialize)]
pub(crate) struct TestZabbixMessage<'a>(pub Vec<TestZabbixValue<'a>>);

impl<'a> Display for TestZabbixMessage<'a> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let json = serde_json::to_string_pretty(self).expect("serialization failed");
        writeln!(f, "{}", json)
    }
}


src/filters.rs
use std::{collections::BTreeMap, convert::Infallible};

use warp::{filters::BoxedFilter, Filter, Rejection};

use crate::handlers::AppContext;

pub fn make_path_filter(path: impl AsRef<str>) -> BoxedFilter<()> {
    let path = path.as_ref().trim_start_matches('/');
    if path.is_empty() {
        warp::path::end().boxed()
    } else {
        let mut segments = path.split('/').map(|s| warp::path(s.to_owned()));
        let first_segment = segments.next().expect("path contains '/'");
        let filter = segments.fold(first_segment.boxed(), |f, s| f.and(s).boxed());
        filter.and(warp::path::end()).boxed()
    }
}

pub fn with_context(ctx: AppContext) -> impl Filter<Extract = (AppContext,), Error = Infallible> + Clone {
    warp::any().map(move || ctx.clone())
}

pub fn get() -> impl Filter<Extract = (crate::JsonValue, ), Error = Rejection> + Clone {
    warp::get()
        // Put GET params into a BTreeMap so they become sorted
        .and(warp::query::<BTreeMap<String, String>>())
        .map(|params| crate::json!(params))
}

pub fn post() -> impl Filter<Extract = (crate::JsonValue, ), Error = Rejection> + Clone {
    warp::post()
        .and(warp::body::json::<crate::JsonValue>())
}


src/logging.rs
use std::collections::HashMap;
use std::fs::File;
use std::marker::PhantomData;
use std::path::{Path, PathBuf};
use std::io::{Write as _, self};

use time::{OffsetDateTime, Instant};
use time::{macros::format_description, format_description::FormatItem};
use tracing::field::Visit;
use tracing::{Subscriber, span, Span};
use tracing_appender::non_blocking::WorkerGuard;
use tracing_subscriber::{
    Layer,
    filter::LevelFilter,
    filter::Targets,
    fmt::{Layer as FmtLayer, MakeWriter},
    prelude::*,
    registry::LookupSpan, layer::Context,
};
use warp::trace::Info;

use crate::AppError;

const CLF_TIME: &[FormatItem] = format_description!("[day]/[month repr:short]/[year]:[hour repr:24]:[minute]:[second] [offset_hour sign:mandatory][offset_minute]");
const HTTP_LOG_TARGET: &str = "warp::filters::trace";

pub(crate) fn setup(
    console_level: LevelFilter,
    access_log: &Option<PathBuf>,
) -> Result<Option<WorkerGuard>, AppError> {
    // Suppress HTTP request logging to console when
    // a file is configured
    let console_log = setup_console(console_level, access_log.is_none());

    let (access_log, worker_guard) = if let Some(file) = access_log {
        let (access_log, worker_guard) = setup_file(file)?;
        (Some(access_log), Some(worker_guard),)
    } else {
        (None, None,)
    };

    tracing_subscriber::registry()
        .with(console_log)
        .with(access_log)
        .try_init()?;
    Ok(worker_guard)
}

fn setup_console<S>(level: LevelFilter, enable_http: bool) -> impl Layer<S>
where
    S: Subscriber + for<'a> LookupSpan<'a>
{
    let main_module = module_path!().split("::").next().unwrap();
    let default_level = level.min(LevelFilter::DEBUG);
    let http_level = if enable_http { default_level } else { LevelFilter::OFF };
    let filters = Targets::default()
        // Don't let the console default level
        // get more granular than Info, because
        // some libraries are VERY verbose (tokio_*)
        .with_default(default_level)
        // Allow this module and zbx_sender to
        // log at Trace and Debug
        .with_target(main_module, level)
        .with_target("zbx_sender", level)
        .with_target("test_mode", LevelFilter::INFO)
        .with_target(HTTP_LOG_TARGET, http_level);
    
    FmtLayer::default()
        .with_filter(filters)
}

fn setup_file<S>(file: &Path) -> Result<(impl Layer<S>, WorkerGuard), io::Error>
where
    S: tracing::Subscriber + for<'lookup> tracing_subscriber::registry::LookupSpan<'lookup>,
{
    let filters = Targets::default()
        // Don't log anything from modules that aren't
        // specified with level_for()
        .with_default(LevelFilter::OFF)
        // HTTP requests must be logged with
        // target = [THIS MODULE]::http
        .with_target(HTTP_LOG_TARGET, LevelFilter::INFO);

    let (log_file, worker_guard) = tracing_appender::non_blocking(File::create(file)?);
    let log_layer = CombinedLogLayer::new(log_file).with_filter(filters);

    Ok((log_layer, worker_guard))
}

struct CombinedLogVisitor {
    start_instant: Instant,
    fields: HashMap<String, String>,
}

impl Default for CombinedLogVisitor {
    fn default() -> Self {
        let start_time = OffsetDateTime::now_local()
            .unwrap_or_else(|_| OffsetDateTime::now_utc());
        let mut fields = HashMap::default();
        fields.insert(
            "start_time".into(),
            start_time.format(CLF_TIME).expect("failed to write timestamp"),
        );

        Self {
            start_instant: Instant::now(),
            fields,
        }
    }
}

impl Visit for CombinedLogVisitor {
    fn record_debug(&mut self, field: &tracing::field::Field, value: &dyn std::fmt::Debug) {
        self.fields.insert(field.to_string(), format!("{:?}", value));
    }

    fn record_str(&mut self, field: &tracing::field::Field, value: &str) {
        self.fields.insert(field.to_string(), value.to_string());
    }
}

impl CombinedLogVisitor {
    fn get_field_value(&self, k: impl AsRef<str>) -> Option<&String> {
        self.fields.get(k.as_ref())
    }
}

struct CombinedLogLayer<
S,
W: for<'writer> MakeWriter<'writer> + 'static,
> {
    make_writer: W,
    _inner: PhantomData<fn(S)>,
}

impl<S, W> CombinedLogLayer<S, W>
where
    W: for<'writer> MakeWriter<'writer> + 'static,
{
    fn new(make_writer: W) -> Self {
        Self {
            make_writer,
            _inner: Default::default(),
        }
    }
}

impl<S, W> Layer<S> for CombinedLogLayer<S, W>
where
    S: tracing::Subscriber + for<'lookup> tracing_subscriber::registry::LookupSpan<'lookup>,
    W: for<'writer> MakeWriter<'writer> + 'static,
{
    fn on_new_span(&self, attrs: &span::Attributes<'_>, id: &span::Id, ctx: Context<'_, S>) {
        let span = ctx.span(id).expect("Span not found, this is a bug");
        let mut extensions = span.extensions_mut();
        if extensions.get_mut::<CombinedLogVisitor>().is_none()
        {
            extensions.insert(CombinedLogVisitor::default());
        }
        let entry = extensions.get_mut::<CombinedLogVisitor>().expect("just inserted");
        attrs.record(entry);
    }

    fn on_event(&self, event: &tracing::Event<'_>, ctx: Context<'_, S>) {
        if let Some(span) = ctx.event_span(event) {
            let mut extensions = span.extensions_mut();
            if let Some(entry) = extensions.get_mut::<CombinedLogVisitor>() {
                event.record(entry);
            }
        }
    }

    fn on_close(&self, id: span::Id, ctx: Context<'_, S>) {
        let span = ctx.span(&id).expect("Span not found, this is a bug");
        let extensions = span.extensions();
        if let Some(entry) = extensions.get::<CombinedLogVisitor>() {
            let mut writer = self.make_writer.make_writer();
            let field_or_blank = |name| {
                entry.get_field_value(name).map_or("-", |s| s.as_str())
            };

            let elapsed = entry.start_instant.elapsed();
            let _ = writeln!(
                writer,
                r#"{} "-" "-" [{}] "{} {} {}" {} 0 "{}" "{}" {:.3}"#,
                field_or_blank("remote.addr"),
                field_or_blank("start_time"),
                field_or_blank("method"),
                field_or_blank("path"),
                field_or_blank("version"),
                field_or_blank("status"),
                field_or_blank("referer"),
                field_or_blank("user_agent"),
                elapsed.as_seconds_f64(),
            );
        }
    }
}

pub fn warp_trace(info: Info<'_>) -> Span {
    use tracing::field::{display, Empty};

    let span = tracing::info_span!(
        target: HTTP_LOG_TARGET,
        "request",
        remote.addr = Empty,
        method = %info.method(),
        path = %info.path(),
        version = ?info.version(),
        referer = Empty,
        user_agent = Empty,
    );

    // Record optional fields.
    if let Some(remote_addr) = info.remote_addr() {
        span.record("remote.addr", &display(remote_addr));
    }

    if let Some(referer) = info.referer() {
        span.record("referer", &display(referer));
    }

    if let Some(user_agent) = info.user_agent() {
        span.record("user_agent", &display(user_agent));
    }

    span
}


src/handlers.rs
use std::{sync::Arc, net::SocketAddr, convert::Infallible};

use thiserror::Error;
use tracing::{info, warn, error};
use warp::{http::StatusCode, Rejection, Reply};

use crate::{AsyncResolver, config::Route, json, JsonValue, zabbix::{ZabbixLogger, TestZabbixValue, TestZabbixMessage}, ZabbixItemValue};

#[derive(Clone)]
pub struct AppContext {
    pub zabbix: Arc<ZabbixLogger>,
    pub resolver: Arc<AsyncResolver>,
    pub test_mode: bool,
}

pub async fn handle_request(
    route: Route,
    ctx: AppContext,
    addr_option: Option<SocketAddr>,
    json: JsonValue,
) -> Result<impl Reply, Rejection> {
    let remote_host = resolve_item_host(&route, &ctx.resolver, &addr_option, &json).await?;

    let payload = match &route.json_filter {
        Some(f) => {
            let jmespath_result = f.search(json).map_err(RequestError::from)?;
            if jmespath_result.is_null() {
                error!("the configured filter [{}] doesn't match any data in the request", f);
                return Err(RequestError::NoFilterMatch.into());
            }
            json!(jmespath_result)
        },
        None => json,
    };

    let item_values = compose_zabbix_kv_tuples(&route.item_key, &payload)?;
    // If running in test mode, log what would be sent to zabbix to the console,
    // and return as if successful.
    if ctx.test_mode {
        let test_values = item_values.iter().map(|i|
            TestZabbixValue {
                host: remote_host.clone(),
                value: i,
            }
        ).collect();
        let test_message = TestZabbixMessage(test_values);

        info!(target: "test_mode", "would send values to Zabbix: {}", test_message);
    } else {
        send_to_zabbix(ctx.zabbix, &remote_host, item_values).await?;
    }

    Ok(warp::reply::with_status(String::from(""), StatusCode::NO_CONTENT))
}

async fn resolve_item_host(
    route: &Route,
    resolver: &AsyncResolver,
    addr_option: &Option<SocketAddr>,
    json: &JsonValue,
) -> Result<String, Rejection> {
    // Is --zabbix-item-host-field specified?
    if let Some(pat) = &route.item_host_field {
        match resolve_dynamic_host(pat, json).await? {
            // Can the requested field be found in the GET params or POST body?
            Some(h) =>
                return Ok(h),
            // Otherwise, abort with an error if --host-field-required is set
            None if route.item_host_field_required =>
                return Err(RequestError::MissingHostField.into()),
            // If --host-field-required not set, then continue below
            None =>
                info!("Specified host field `{}` is not present", pat),
        }
    }

    if let Some(host) = &route.item_host {
        // Is a static --zabbix-item-host set?
        Ok(host.to_owned())
    } else if let Some(addr) = addr_option {
        // Otherwise, try to resolve the client's address to
        // a DNS name and fall back to just using the IP
        let addr = addr.ip();
        let lookup_result = resolver.reverse_lookup(addr).await;
        let host_ptr = lookup_result
            // Convert Ok into Some(PTR result), Err into None
            .iter()
            // Create an iterator over reverse_lookup results,
            // and flatten with above Iter over Result
            .flat_map(|r| r.iter())
            // Convert result to string, and trim trailing '.'
            .map(|v| v.to_string().trim_end_matches('.').to_owned())
            // Get the first result (should be only one for PTR records)
            .next();

        // Use Some(PTR result) or fall back to the IP address as String
        let host = host_ptr.unwrap_or_else(|| addr.to_string());

        Ok(host)
    } else {
        // Unlikely, but if all ways of determining the item
        // host are unavailable, abort with an error
        Err(RequestError::MissingClientAddr.into())
    }
}

async fn resolve_dynamic_host(
    pattern: &jmespath::Expression<'_>,
    json: &JsonValue,
) -> Result<Option<String>, RequestError> {
    let field = pattern.search(json)?;

    // Ensure result of JMESPath query is a non-empty string
    let field = match field.as_string() {
        Some(s) if s.is_empty() => {
            warn!("Specified host field `{}` is present but is an empty string", pattern);
            None
        },
        Some(s) => Some(s.to_owned()),
        None => {
            // Warn on wrong field type, but quietly pass along None for null
            if !field.is_null() {
                warn!("Specified host field `{}` is present but is not a string (found type {})", pattern, field.get_type());
            }
            None
        }
    };
    Ok(field)
}

fn compose_zabbix_kv_tuples(key: &str, data: &JsonValue) -> Result<Vec<ZabbixItemValue>, RequestError> {
    let values = if key == "*" {
        data.as_object()
            .ok_or(RequestError::WildcardKeyOnNonObject)?
            .into_iter()
            .map(|v| ZabbixItemValue {
                key: v.0.to_string(),
                value: stringify_zabbix_value(v.1),
            })
            .collect()
    } else {
        vec![ZabbixItemValue {
            key: key.to_owned(),
            value: data.to_string(),
        }]
    };
    Ok(values)
}

async fn send_to_zabbix(zabbix: Arc<ZabbixLogger>, host: &str, values: Vec<ZabbixItemValue>) -> Result<(), RequestError> {
    let host = host.to_owned();
    let result = tokio::task::spawn_blocking(move || zabbix.log_many(&host, values)).await?;
    match result {
        Ok(res) => {
            match res.failed_cnt() {
                None => Err(RequestError::ZabbixBadReply),
                Some(n) if n > 0 => Err(RequestError::ZabbixItemsFailed(n)),
                _ => Ok(()),
            }
        }
        Err(err) => {
            error!("Zabbix error: {}", err);
            Err(RequestError::ZabbixError(err.to_string()))
        },
    }
}

fn stringify_zabbix_value(value: &JsonValue) -> String {
    if let Some(s) = value.as_str() {
        s.to_string()
    } else if let Some(b) = value.as_bool() {
        let bool_as_int_str = if b { "1" } else { "0" };
        bool_as_int_str.to_string()
    } else {
        value.to_string()
    }
}

pub async fn handle_errors(err: Rejection) -> Result<impl Reply, Infallible> {
    let code;
    let message;

    if err.is_not_found() {
        code = StatusCode::NOT_FOUND;
        message = "NOT_FOUND";
    } else if let Some(e) = err.find::<RequestError>() {
        use RequestError::*;
        (code, message) = match e {
            MissingClientAddr => (StatusCode::BAD_REQUEST, "MISSING_CLIENT_ADDR",),
            JmespathError(_) => (StatusCode::INTERNAL_SERVER_ERROR, "JMESPATH_ERROR",),
            MissingHostField => (StatusCode::BAD_REQUEST, "MISSING_HOST_FIELD",),
            NoFilterMatch => (StatusCode::BAD_REQUEST, "MISSING_FILTER_DATA",),
            JoinError(_) => (StatusCode::INTERNAL_SERVER_ERROR, "ZABBIX_TASK_FAILED",),
            ZabbixBadReply => (StatusCode::INTERNAL_SERVER_ERROR, "ZABBIX_REPLY_INVALID",),
            ZabbixItemsFailed(_) => (StatusCode::BAD_REQUEST, "ZABBIX_ITEMS_FAILED",),
            ZabbixError(_) => (StatusCode::BAD_GATEWAY, "ZABBIX_ERROR",),
            WildcardKeyOnNonObject => (StatusCode::BAD_REQUEST, "DATA_NOT_AN_OBJECT",),
        };
    } else {
        code = StatusCode::INTERNAL_SERVER_ERROR;
        message = "UNHANDLED_ERROR";
    }

    let json = warp::reply::json(&json!({
        "code": code.as_u16(),
        "message": message,
    }));

    Ok(warp::reply::with_status(json, code))
}

#[derive(Error, Debug)]
enum RequestError {
    #[error("the client's remote address was not available")]
    MissingClientAddr,
    #[error(transparent)]
    JmespathError(#[from] jmespath::JmespathError),
    #[error("the request parameters do not contain the host field")]
    MissingHostField,
    #[error("the supplied filter did not return any data from the request")]
    NoFilterMatch,
    #[error("the item key was specified as a wildcard ('*') but the supplied item value was not a JSON object")]
    WildcardKeyOnNonObject,
    #[error("Zabbix trapper task failed: {0}")]
    JoinError(#[from] tokio::task::JoinError),
    #[error("Zabbix returned a non-number where the failed count should have been")]
    ZabbixBadReply,
    #[error("Zabbix failed {0} items in the request")]
    ZabbixItemsFailed(i32),
    #[error("{0}")]
    ZabbixError(String),
}

impl warp::reject::Reject for RequestError {}


src/config.rs
use std::{net::{SocketAddr, Ipv6Addr}, path::PathBuf, str::FromStr, fs};

use clap::Parser;
use tracing_subscriber::filter::LevelFilter;
use serde::{Deserialize, Deserializer, de};
use warp::http::uri::PathAndQuery;

use crate::AppError;

type JmesPathExpression = jmespath::Expression<'static>;

const DEFAULT_CONFIG_PATH: &str = concat!("/etc/", env!("CARGO_PKG_NAME"), ".toml");

#[derive(Parser, Clone, Debug)]
#[clap(about, author)]
struct Cli {
    /// Path to TOML-format configuration file
    #[clap(long = "config", short = 'c', parse(from_os_str), default_value = DEFAULT_CONFIG_PATH)]
    config_file: PathBuf,

    #[clap(long = "listen", short = 'l')]
    /// HTTP server listening address and port
    ///
    /// [default: 3030]
    listen: Option<SocketAddr>,

    #[clap(long = "zabbix-server", short = 'z', display_order(1))]
    /// Zabbix Server address
    zabbix_server: Option<String>,

    #[clap(long = "zabbix-port", short = 'p', display_order(2))]
    /// Zabbix Server trapper port
    ///
    /// [default: 10051]
    zabbix_port: Option<u16>,

    #[clap(long = "access-log", parse(from_os_str))]
    /// Log to a file in Apache Combined logging format
    access_log_path: Option<PathBuf>,

    #[clap(short, parse(from_occurrences = Cli::parse_verbosity))]
    /// Specify up to 3 times to increase console logging
    verbosity: LevelFilter,

    #[clap(long)]
    /// Run in test mode without sending actual values to Zabbix server
    test_mode: bool,
}

impl Cli {
    fn parse_verbosity(count: u64) -> LevelFilter {
        match count {
            0 => LevelFilter::OFF,
            1 => LevelFilter::INFO,
            2 => LevelFilter::DEBUG,
            _ => LevelFilter::TRACE,
        }
    }
}

#[derive(Deserialize, Clone)]
 pub struct Config {
    #[serde(default = "Config::default_listen")]
    /// HTTP server listening address and port
    pub listen: SocketAddr,

    /// Zabbix Server address
    pub zabbix_server: Option<String>,

    #[serde(default = "Config::default_zabbix_port")]
    /// Zabbix Server trapper port
    pub zabbix_port: u16,

    /// Log to a file in Apache Combined logging format
    pub access_log_path: Option<PathBuf>,

    #[serde(deserialize_with = "Config::parse_log_level", default = "Config::default_log_level")]
    pub log_level: LevelFilter,

    // Rename to singular to improve config file legibility
    #[serde(rename = "route")]
    pub routes: Vec<Route>,

    #[serde(skip, default)]
    pub test_mode: bool,
}

impl Config {
    fn default_listen() -> SocketAddr { (Ipv6Addr::LOCALHOST, 3030).into() }
    fn default_zabbix_port() -> u16 { 10051 }
    fn default_log_level() -> LevelFilter { LevelFilter::WARN }
    fn parse_log_level<'de, D>(deserializer: D) -> Result<LevelFilter, D::Error>
        where D: Deserializer<'de>
    {
        let s = String::deserialize(deserializer)?;
        LevelFilter::from_str(&s)
            .map_err(|_| de::Error::custom("log_level value doesn't match any known log level"))
    }

    pub(crate) fn load() -> Result<Self, AppError> {
        let args = Cli::parse();
        let toml_file = fs::read_to_string(&args.config_file)
            .map_err(|e| AppError::ConfigLoadError {
                path: args.config_file.to_string_lossy().into_owned(),
                source: e
            })?;
        let mut config: Self = toml::from_str(&toml_file)?;
        config.listen = args.listen.unwrap_or(config.listen);
        config.zabbix_server = args.zabbix_server.or(config.zabbix_server)
            .ok_or_else(|| AppError::ConfigMissingRequired("zabbix_server".into()))?
            .into();
        config.zabbix_port = args.zabbix_port.unwrap_or(config.zabbix_port);
        config.log_level = args.verbosity.max(config.log_level);
        config.access_log_path = args.access_log_path.or(config.access_log_path);
        config.test_mode = args.test_mode;
        Ok(config)
    }
}

#[derive(Deserialize, Clone)]
pub struct Route {
    #[serde(deserialize_with = "Route::parse_path")]
    /// Path on which to accept requests
    pub path: PathAndQuery,

    /// Host name for Zabbix Item (OPTIONAL) [default: reverse DNS or IP address of HTTP client]
    ///
    /// Host name the item belongs to (as registered in Zabbix frontend).
    pub item_host: Option<String>,

    // Using deserialize_with on an Option<T> means that default has to be
    // specified, or a missing field becomes an error. See
    // https://github.com/serde-rs/serde/issues/723#issuecomment-423299411
    #[serde(default, deserialize_with = "Route::parse_option_jmespath")]
    /// Dynamic field from request to determine host name for Zabbix Item.
    /// Can be a simple top-level field name or a JMESpath filter. In POSTed
    /// JSON data, the result must be a JSON string.
    /// [default: specified item_host or HTTP client address]
    pub item_host_field: Option<JmesPathExpression>,

    #[serde(default)]
    /// The field specified by item_host_field must be present in the request, or
    /// a warning will be logged and the request dropped.
    pub item_host_field_required: bool,

    /// Zabbix Item key. The special value "*" means expand all top
    /// level keys in the request data to individual Zabbix item keys.
    pub item_key: String,

    // See comment above about "redundant" serde(default)
    #[serde(default, deserialize_with = "Route::parse_option_jmespath")]
    /// JMESpath filter to be applied to data before forwarding to Zabbix.
    /// GET parameters are transformed from key=value... to {"key": "value",...}
    /// and can be filtered as normal JSON.
    /// POSTed JSON data can be filtered directly.
    pub json_filter: Option<JmesPathExpression>,
}

impl Route {
    fn parse_path<'de, D>(deserializer: D) -> Result<PathAndQuery, D::Error>
        where D: Deserializer<'de>
    {
        let s = String::deserialize(deserializer)?;
        PathAndQuery::from_str(&s) 
            .map_err(de::Error::custom)
            .and_then(|path| match path.query() {
                Some(q) => Err(de::Error::custom(format!("route path may not contain the query string `{}`", q))),
                None => Ok(path),
            })
    }

    fn parse_option_jmespath<'de, D>(deserializer: D) -> Result<Option<JmesPathExpression>, D::Error>
        where D: Deserializer<'de>
    {
        let maybe_s: Option<String> = Option::deserialize(deserializer)?;
        let maybe_jmespath = match maybe_s {
            None => None,
            Some(s) => {
                let j = jmespath::compile(&s)
                    .map_err(de::Error::custom)?;
                Some(j)
            },
        };
        Ok(maybe_jmespath)
    }
}
